File: frontend/lib/notesApiServices/Semesters/semester.dart
class Semester {
  final String? id;
  final String userId;
  final String title;
  final DateTime createdAt;

  Semester({
    this.id,
    required this.userId,
    required this.title,
    required this.createdAt,
  });

  factory Semester.fromJson(Map<String, dynamic> json) {
    return Semester(
      id: json['id'],
      userId: json['userId'],
      title: json['title'],
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "userId": userId,
      "title": title,
      "createdAt": createdAt.toIso8601String(),
    };
  }
}

File: backend/models/Semester.js

import mongoose from "mongoose";
import Subject from "./Subject.js"; // Make sure path is correct

const semesterSchema = new mongoose.Schema({
  userId: { type: String, required: true },
  title: { type: String, required: true },
  createdAt: { type: Date, default: Date.now }
});

// Return `id` instead of `_id` in JSON
semesterSchema.set('toJSON', {
  virtuals: true,
  versionKey: false,
  transform: (doc, ret) => { delete ret._id; }
});

/**
 * Pre-remove middleware to handle cascading deletes
 */
semesterSchema.pre("deleteOne", { document: true, query: false }, async function(next) {
  try {
    const subjects = await Subject.find({ semesterId: this.id });
    // for (let subject of subjects) {
    //   await subject.deleteOne(); // recursive delete
    // }
    await Promise.all(subjects.map(subject => subject.deleteOne()));  // 🔹 parallel delete

    next();
  } catch (err) {
    next(err);
  }
});

const Semester = mongoose.model("Semester", semesterSchema);

export default Semester;
